design/peer-awareness.md

```diff
@@ -0,0 +1,95 @@
+---
+title: peer-awareness
+target-version: 0.1.2
+---
+
+# Peer Awareness
+
+## Summary
+
+We will provide Peer Awareness which is a simple algorithm that manages end-user
+status like who is connected and metadata like username or email address and etc.
+For example, users can implement a list of people participating in the editing,
+such as a box in the top right of Google Docs.
+
+### Goals
+
+Implement Peer Awareness and provide API to users to use the feature. The goal of
+the first version is to implement simple functionality and check usability.
+
+### Non-Goals
+
+The first version does not implement complex features such as dynamic metadata
+updates.
+
+## Proposal details
+
+### How to use
+
+Users can pass metadata along with client options when creating a client.
+
+```typescript
+const client = yorkie.createClient('https://yorkie.dev/api', {
+  metadata: {
+    username: 'hackerwins'
+  }
+});
+```
+
+Then the users create a document in the usual way and attach it to the client.
+
+```typescript
+const doc = yorkie.createDocument('examples', 'codemirror');
+await client.attach(doc);
+```
+
+When a new peer registers or leaves, `peers-changed` event is fired, and the
+other peer's clientID and metadata can be obtained from the event.
+
+```typescript
+client.subscribe((event) => {
+  if (event.name === 'peers-changed') {
+    const peers = event.value[doc.getKey().toIDString()];
+    for (const [clientID, metadata] of Object.entries(peers)) {
+      console.log(clientID, metadata);
+    }
+  }
+});
+```
+
+### How does it work?
+
+```
+ +--Client "A"----+                      +--Agent-----------------+
+ | +--Metadata--+ |                      | +--PubSub -----------+ |
+ | | { num: 1 } | <-- WatchDocuments --> | | {                  | |
+ | +------------+ |                      | |   docA: {          | |
+ +----------------+                      | |     A: { num: 1 }, | |
+                                         | |     B: { num: 2 }  | |
+ +--Client "B"----+                      | |   },               | |
+ | +--Metadata--+ |                      | |   ...              | |
+ | | { num: 2 } | <-- WatchDocuments --> | | }                  | |
+ | +------------+ |                      | +--------------------+ |
+ +----------------+                      +------------------------+
+```
```

Comment:
I think we need to turn a diagram written in text like this into an image file. But I don't think it is now. When the branding of our project is done, it looks like we can draw it by applying our colors, designs, and feelings.
This can be a cumbersome task, but it seems that even if there are more documents, you can have unity. What do you think?

Reply:
Thanks for the comment. In my opinion, there seems to be no need to unify the style of diagrams in internal documents. However, it would be nice to apply it to the documentation we write on https://yorkie.dev.